Filter optimizer logical op "or" and "one"

stevenyoungs opened the pull request:
Enable the filter optimizer to optimize filters the use logical operations of "one" or "or"

Note: this branch needs to be rebased once #2052 is merged into maintenance/gramps60 and onward merged into master

stevenyoungs committed:
The Optimizer class should use the parent filters logical_op rather than the sub-filters logical_op when calculating the result.

Fixes #13799

stevenyoungs committed:
Add a unit test of the filter optimizer

Co-authored-by: Doug Blank <doug.blank@gmail.com>

dsblank committed:
Added new CustomFilters.remove(); rearrange optimizer tests

dsblank committed:
Don't assume namespace exists in remove()

dsblank committed:
Add copyright

stevenyoungs committed:
Add new files to POTFILES.skip

dsblank committed:
Added second issue test

dsblank committed:
Linting

stevenyoungs committed:
Log additional information in the Optimizer class

dsblank committed:
Should pass the parent logical op?

stevenyoungs committed:
reimplement Optimizer

Co-authored-by: Kari Kujansuu <kari.kujansuu@gmail.com>

stevenyoungs committed:
When logical_op is "one" we can use symmetric_difference rather than union to further reduce the search space

stevenyoungs committed:
Load custom filters directly from a string

stevenyoungs committed:
testing

stevenyoungs committed:
testing a bit more

stevenyoungs committed:
testing a bit less

dsblank committed:
Attempt to fix failing other filters

stevenyoungs committed:
Use union rather than symmetric_difference when logical_op is "one"

selected_handles \ all_handles is a superset of the potential result of a rule. Therefore symmetric_difference cannot be used reliably - it will potentially exclude handles which should be in the result.

stevenyoungs committed:
Add docstring

stevenyoungs committed:
Simplify the implementation of intersection and union

stevenyoungs committed:
Improve comments

stevenyoungs committed:
handles_in is never None

stevenyoungs committed:
remove handles_out as it is no longer used

stevenyoungs committed:
lint

stevenyoungs committed:
Optimize intersection to avoid computing x&x when we have a single set

stevenyoungs committed:
Improve comments in Optimizer

stevenyoungs committed:
rename handles_in to possible_handles to better convey the intent

stevenyoungs committed:
Further refinement of comments

stevenyoungs committed:
Add additional filter optimizer tests

stevenyoungs committed:
Fix test result

stevenyoungs committed:
Test permutations of empty filters

stevenyoungs committed:
Fix expected test result

dsblank committed:
Optimize intersection

dsblank committed:
Optimize union

dsblank committed:
Optimize intersection

dsblank committed:
Optimize when you start with id_list

dsblank committed:
testing: show reason if failure

stevenyoungs committed:
Refactor and simplify the code to handle id_list and id_list == None in the same control flow

dsblank committed:
testing: show reason if failure

stevenyoungs committed:
Move the check for None handles to after the Optimizer

stevenyoungs committed:
Rearrange code

stevenyoungs committed:
Use the correct starting set of possible_handles when tree == True

stevenyoungs committed:
intersection must start with sets[0].

If we start with an empty set, the overall result, by definition, will always be empty.

stevenyoungs committed:
The optimisation for the case when there is only 1 set is no longer required.

When len(sets) == 1, the initial value of reduce will be sets[0], the iterable will be None and there will be no intersection calculated

stevenyoungs committed:
lint

stevenyoungs committed:
When tree==True, the order of final_list must match the order the handles are traversed by get_tree_cursor

stevenyoungs committed:
when id_list is a list of tuples, we need to return the matching tuples, not the matching handles

stevenyoungs committed:
Add more detail to the docstring for apply

stevenyoungs committed:
Add type hinting

stevenyoungs committed:
Simplify the lower level apply_logical_op_to_all method by moving the complexities of determing the initial set of possible_handles into apply, along with converting the result handles back in to the correct type

stevenyoungs committed:
Fix typing error

stevenyoungs committed:
Fix typing

stevenyoungs committed:
Fix typing

stevenyoungs committed:
fix typing

stevenyoungs committed:
Avoid adding None to self.selected_handles

dsblank committed:
Added tests for id_list, tupleind, and tree

dsblank committed:
No pytest

dsblank committed:
Use assertRaises

stevenyoungs committed:
Use set.intersection instead of reduce

Co-authored-by: Douglas Blank <doug.blank@gmail.com>

stevenyoungs committed:
remove import that is no longer required

stevenyoungs committed:
maintain the input order in the result when id_list and tupleind are provided

stevenyoungs committed:
minor simplification of code and clarify comment

stevenyoungs committed:
Fix spelling in comment

stevenyoungs committed:
Clarify comment

stevenyoungs committed:
Improve comment

stevenyoungs committed:
Check is not None

stevenyoungs committed:
Add typing to Optimizer

stevenyoungs committed:
lint

stevenyoungs committed:
Break circular dependency in imports unless type checking

stevenyoungs committed:
linting

stevenyoungs committed:
Add additional filter optimizer with inverted result and a non-optimized rule

stevenyoungs committed:
Delete extra copy-and-paste from previous commit

stevenyoungs committed:
When a filter is inverted, the default set of potential handles matching a rule is the empty set. Later, when we computer the difference(set(), self.all_handles) the correct result of self.all_handles is obtained.

dsblank committed:
Add failing not family test

dsblank committed:
Refactor proper use of evidence

dsblank committed:
Linting

dsblank committed:
More liniting

dsblank committed:
Update gramps/gen/filters/optimizer.py

Co-authored-by: stevenyoungs <steve@youngs.cc>

dsblank committed:
Update gramps/gen/filters/optimizer.py

Co-authored-by: stevenyoungs <steve@youngs.cc>

dsblank committed:
Update gramps/gen/filters/optimizer.py

Co-authored-by: stevenyoungs <steve@youngs.cc>

dsblank committed:
Update gramps/gen/filters/optimizer.py

Co-authored-by: stevenyoungs <steve@youngs.cc>

dsblank committed:
Liniting of suggestions

dsblank committed:
Update gramps/gen/filters/optimizer.py

dsblank committed:
Use intersection for AND

dsblank committed:
Optimizer last fix :)

dsblank commented:
The only way we can optimize "or" would be if all rules (at that level) have selected_handles.

stevenyoungs commented:
> The only way we can optimize "or" would be if all rules (at that level) have selected_handles.

Yes I reached that conclusion as well and gave up for a bit. It may just be inverted unoptimized rules that are a problem when the logical op is "or". Let's see if my latest attempt passes the tests.....

stevenyoungs committed:
Add missing filters "I0001" and "I0002" which are used by the filter "not (I0001 and I0002)"

dsblank committed:
Added single tests for individuals

dsblank committed:
Added single tests for individuals

stevenyoungs committed:
Add type hints to GenericFilter.logical_op

stevenyoungs committed:
Support optimization of filters using logical operations "one" and "or"

If there is more than one rule, for "one" or "or", then all rules must be optimized in order to optimize the filter.

stevenyoungs committed:
Remove unused code

stevenyoungs committed:
Add test using an optimized rule ("parents of") along with a filter that uses "or" and invert

stevenyoungs committed:
Disable support for optimizing filters with logical_op "one" or "or"

stevenyoungs committed:
Enable optimization of filters where the logical_op is "or" or "one"
